### PR TITLE
Store FramePointer in IR

### LINKED ISSUE #10129: Store FramePointer in the IR and cache it
This inquiry is inspired by this [IrToTruffle code](https://github.com/enso-org/enso/blob/e9452ea8ad3bf6fc3216d785117b554a8318dbf0/engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala#L1685). First of all it, takes `Occurance` from the IR:
```java
          val useInfo = name
            .unsafeGetMetadata(
              AliasAnalysis,
              "No occurrence on variable usage."
            )
            .unsafeAs[AliasInfo.Occurrence]

          val framePointer = scope.getFramePointer(useInfo.id)
```
and it converts it into `FramePointer` to do that it needs a `scope` also coming from the IR (some parent).  The conversion of `Occurence` via `LocalScope` is non-trivial as observed by 
- #9235

However, **we don't have to do it at all**! All we need is the `FramePointer` (a pair of integers), we don't need all the scopes & co. As such:

Let's compute `FramePointer` onces, before **generating caches** and then just load it.

```[tasklist]
### Tasks
- [ ] Implement an `IRPass` to compute the `FramePointer` and store it as _metadata_
- [ ] Change `IrToTruffle` to just read `FramePointer`
- [ ] Ideally make sure `AliasInfo` & co. isn't read from caches at all
- [ ] Verify some benchmark times have improved
```

### PR DESCRIPTION
**Pull request by Akirathan**
Fixes #10129

### Pull Request Description

Adds a new compiler pass [FramePointerAnalysis](https://github.com/enso-org/enso/blob/af1b7e8d97032628a7276821f9e3b5cdfcc1165c/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala) that attaches [FramePointer]https://github.com/enso-org/enso/blob/af1b7e8d97032628a7276821f9e3b5cdfcc1165c/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala#L386) metadata to various IR elements.

FramePointer metadata serves as a replacement for usage of `AliasMetadata.Occurence` in `IrToTruffle`.

There are still usages of `AliasAnalysis` metadata from `IrToTruffle`. Which means that alias metadata is still deserialized in `IrToTruffle`, and therefore, we will not see any big startup time improvements. To remove all usages of `AliasAnalysis` metadata in `IrToTruffle`, we need to think about how to remove all alias metadata references from [LocalScope](https://github.com/enso-org/enso/blob/9f4e71ffdf03e6777eb212206564f8ffb577b607/engine/runtime-compiler/src/main/scala/org/enso/compiler/context/LocalScope.scala) which seems like a big task for a follow-up PR.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+import org.enso.compiler.core.ir.Module
+import org.enso.compiler.core.ir.{Expression, ProcessingPass}
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.IRPass.IRMetadata
+
+/** This pass attaches [[FramePointer]] as metadata to all the IR elements that already
+  * have [[org.enso.compiler.pass.analyse.alias.Info.Occurrence]] attached.
+  */
+case object FramePointerAnalysis extends IRPass {
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala:34]: A pass in Scala, eh!

File: engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/FramePointerAnalysisTest.scala [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    "attach frame pointers to local variables" in {
+      implicit val ctx: ModuleContext = mkModuleContext
+      val ir =
+        """
+          |main =
+          |    a = 1
+          |    b = 2
+          |    42
+          |""".stripMargin.preprocessModule.analyse
+      val allOcc = collectAllOccurences(ir)
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/analyse/FramePointerAnalysisTest.scala:73]: A reasonable _generic test_ is to verify that each IR that has occurrence also has a `FramePointer`.

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+  }
+
+  private def processBinding(
+    ir: Definition
+  ): Definition = {
+    ir match {
+      case m: Method.Explicit =>
+        getAliasAnalysisGraph(m) match {
+          case Some(graph) =>
+            m.copy(
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: Why do we do `copy` when this pass is only supposed to update `Metadata` and that can be done _in place_ inside of the original `IR`?

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+              body = m.body.mapExpressions(processExpression(_, graph))
+            )
+          case _ => m
+        }
+      case t: Definition.Type =>
+        getAliasAnalysisGraph(t) match {
+          case Some(graph) =>
+            t.copy(
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: - btw. see @hubertp #10718 to see most recent thoughts about minimizing copying

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+                val frameSlotIdx = getFrameSlotIdxInScope(graph, scope, defn)
+                ir.updateMetadata(
+                  new MetadataPair(
+                    this,
+                    new FramePointerMeta(
+                      new FramePointer(parentLevel, frameSlotIdx)
+                    )
+                  )
+                )
+              case _ => ir
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: We seem to do a lot of `.copy` in other methods, but here, at the core method, we just _update the IR_ and return the same `ir` reference back. This makes me believe all the copying is unnecessary.

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.IRPass.IRMetadata
+import org.enso.compiler.pass.analyse.alias.{Graph, Info}
+
+/** This pass attaches [[FramePointer]] as metadata to all the IR elements that already
+  * have [[org.enso.compiler.pass.analyse.alias.Info.Occurrence]] attached.
+  */
+case object FramePointerAnalysis extends IRPass {
+
+  override type Metadata = FramePointerMeta
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala:36]: `FramePointerMeta` seems to add one additional and unnecessary level of indirection. It'd be simpler if the object in the `Metadata` was directly `FramePointer`.

Refactoring of [various IRMetadata concepts](https://github.com/enso-org/enso/pull/10729/files#diff-bd9022d776b808f603e287826601f2d1dc9bf703a17972645ebb1ff7ab7de7ebR330) would however be very likely needed...

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala (hunk 3 above)
Comment (Akirathan) [reply engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: Good point. Removed all the copying in 61821bf3932f72bac1355f2aed3c05af6972d02b

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala (hunk 4 above)
Comment (Akirathan) [reply engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: Copying removed in 61821bf3932f72bac1355f2aed3c05af6972d02b

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala (hunk 5 above)
Comment (Akirathan) [reply engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala]: All copying removed in 61821bf3932f72bac1355f2aed3c05af6972d02b

**Comment by JaroslavTulach** [comment]
I don't see changes in `IrToTruffle` to use these new metadata. Those are essential in verifying the PR changes are correct.

**Comment by JaroslavTulach** [comment]
What does it mean?
```
org.graalvm.polyglot.PolyglotException: java.lang.UnsupportedOperationException: unimplemented, took 0.34 sec
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.compiler.pass.analyse.types.TypePropagation$CompilerNameResolution.assertLinkEquals(TypePropagation.java:394)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.compiler.pass.analyse.types.TypePropagation$CompilerNameResolution.assertLinkEquals(TypePropagation.java:336)
 INFO ide_ci::program::command: sbt ℹ️ [error]     at org.enso.compiler.context.NameResolutionAlgorithm.resolveName(NameResolutionAlgorithm.java:47)
```
at https://github.com/enso-org/enso/actions/runs/10373180715/job/28717694507?pr=10729

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Store FramePointer in IR · enso-org/enso@6ec1ba6</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/context/LocalScope.scala [hunk 7]
```diff
@@ -73,43 +73,6 @@ class LocalScope(
...
-  }
-
-  /** Obtains the frame pointer for a given identifier from the current scope, or from
-    * any parent scopes.
-    *
-    * @param id the identifier of a variable usage occurrence from alias
-    *           analysis
-    * @return the frame pointer for `id`, if it exists
-    */
-  def getFramePointer(id: AliasGraph.Id): Option[FramePointer] = {
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/context/LocalScope.scala:99]: We no longer need `LocalScope` to compute `getFramePointer`. That's good.

**Review by JaroslavTulach — APPROVED** [review]
The changes in `IrToTruffle` are good. We no longer need `AliasAnalysis` for computing `FramePointer`. Let's get this in.

However, as the primary goal was to speed things up, we need a away to create `LocalScope` without reading the alias graphs. Can we use `Persistance.Reference` there?
```diff
--- engine/runtime-compiler/src/main/scala/org/enso/compiler/context/LocalScope.scala
+++ engine/runtime-compiler/src/main/scala/org/enso/compiler/context/LocalScope.scala
@@ -31,8 +31,8 @@ import scala.jdk.CollectionConverters._
   */
 class LocalScope(
   final val parentScope: Option[LocalScope],
-  final val aliasingGraph: AliasGraph,
-  final val scope: AliasGraph.Scope,
+  final val aliasingGraph: Persistance.Reference[AliasGraph],
+  final val scope: Persistance.Reference[AliasGraph.Scope],
   final val dataflowInfo: DataflowAnalysis.Metadata,
   final val flattenToParent: Boolean                       = false,
   private val parentFrameSlotIdxs: Map[AliasGraph.Id, Int] = Map()
```

**Comment by Akirathan** [comment]
Follow-up task created in https://github.com/enso-org/enso/issues/10833

**Pull request merged**